File: 2 pointers/42_trapping_rain_water.ts
function trap(height: number[]): number {


    let N = height.length, ans = 0;
    let bigL = new Array(N).fill(0), bigR = new Array(N).fill(0);

    let prev = 0;
    for(let i=0; i<N; i++){
        bigL[i] = Math.max(prev, height[i]);
        prev = bigL[i];
    }

    prev = 0;
    for(let i=N-1; i>-1; i--){
        bigR[i] = Math.max(prev, height[i]);
        prev = bigR[i];
    }

    for(let i=0; i<N; i++){
        ans += (Math.min(bigL[i], bigR[i]) - height[i]);
    }

    return ans;

};

File: 2 pointers/42_trapping_rain_water.rs
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {

        let N:usize = height.len();
        let mut big_l:Vec<i32> = vec![0; N];
        let mut big_r:Vec<i32> = vec![0; N];
        let mut ans:i32 = 0;

        let mut prev:i32 = 0;
        for i in 0.. N {
            big_l[i] = i32::max(height[i], prev);
            prev = big_l[i];
        }

        prev = 0;
        for i in (0.. N).rev() {
            big_r[i] = i32::max(height[i], prev);
            prev = big_r[i];
        }

        for i in 0.. N {
            ans += (i32::min(big_l[i], big_r[i]) - height[i]);
        }

        ans

    }
}